exonum/src/blockchain/tests.rs

```diff
@@ -361,6 +361,96 @@ mod transactions_tests {
     }
 }
 
+struct ServiceGood;
+
+impl Service for ServiceGood {
+    fn service_id(&self) -> u16 {
+        1
+    }
+
+    fn service_name(&self) -> &'static str {
+        "some_service"
+    }
+
+    fn state_hash(&self, _snapshot: &Snapshot) -> Vec<Hash> {
+        vec![]
+    }
+
+    fn tx_from_raw(&self, _raw: RawTransaction) -> Result<Box<Transaction>, MessageError> {
+        unimplemented!()
+    }
+
+    fn execute(&self, fork: &mut Fork) {
+        let mut index = ListIndex::new(IDX_NAME, fork);
+        index.push(1);
+    }
+}
+
+struct ServicePanic;
+
+impl Service for ServicePanic {
+    fn service_id(&self) -> u16 {
+        1
+    }
+
+    fn service_name(&self) -> &'static str {
+        "some_service"
+    }
+
+    fn state_hash(&self, _snapshot: &Snapshot) -> Vec<Hash> {
+        vec![]
+    }
+
+    fn tx_from_raw(&self, _raw: RawTransaction) -> Result<Box<Transaction>, MessageError> {
+        unimplemented!()
+    }
+
+    fn execute(&self, _fork: &mut Fork) {
+        panic!("42");
```

Comment:
If I understand the logic of tests correctly, this method should write to the index and *then* panic; then corresponding tests would actually check that the corresponding changes are actually rolled back.

Reply:
> and what about assert!(index.is_empty()); at line 451 ?

The index should still be empty because of rollback, shouldn't it?

Reply:
Exactly